website/photos/services.py

```diff
@@ -70,3 +84,99 @@ def get_annotated_accessible_albums(request, albums):
         )
 
     return albums
+
+
+def extract_archive(request, album, archive):
+    iszipfile = is_zipfile(archive)
+    archive.seek(0)
+
+    if iszipfile:
+        with ZipFile(archive) as zip_file:
+            for photo in sorted(zip_file.infolist(), key=lambda x: x.filename):
+                extract_photo(request, zip_file, photo, album)
+    else:
+        # is_tarfile only supports filenames, so we cannot use that
+        try:
+            with tarfile.open(fileobj=archive) as tar_file:
+                for photo in sorted(tar_file.getmembers(), key=lambda x: x.name):
```

Comment:
I believe this is very inefficient for tar files because they don't have a directory like zip files do. In zip files you can read the file listing from the start of the file but for tar.{gz,bzip} you have to decompress and scan through the file.

Reply:
Probably true, but based on the fact that this is only moving around code that already existed (https://github.com/svthalia/concrexit/pull/1059/files/ff339ac8d255ead53fef8dd406f29d58b5b8eccf#diff-a75c0fdbb5737477505a8678ac15706fL126) I think optimalisations to the code can be moved to a new issue